[2240] - Robonect Binding

reyem opened the pull request:
Implemented features:
* retrieving status and version information
* change mode (AUTO, MANUAL, HOME )
* update name 
* Define and start jobs
* Optional authentication settings.
tested against robonect firmware 0.9c

Open/Limitations:
* Jobs API not fully understood and therefore most probably buggy.

itheiss commented:
Great that you took the challenge to develop an openHAB2 Binding! Thanks!

As soon as authentication is working I will start testing and give you feedback.

reyem commented:
@itheiss Thank you! Great! I committed the authentication bit. So when adding the thing you can optionally pass username and password as well as the poll interval. On my mower it looked so far ok, although I haven't done extensive tests yet...

idznak commented:
Hi i've just installed your Robonect Binding via copying in the addons folder. 
In OH2 - PaperUI i' tried to add a new "Things" by choose Binding "Robonect Binding" but nothing was shown...
i'have a Gardena R40LI with a Robonect Hx since yesterday and OH2 since today

reyem commented:
Hi idznak,
After chooding the binding you should be presented with a mask, where you can enter the ip etc.

Did you install OH 2.0 release or 2.1-Snapshot?
The binding is built againdt the snapshot release and I'm not sure if it runs with the stable release. So please use the latest oh 2.1-Snapshot.

Also do you have any error logs?

idznak commented:
Strange... now it's there. Seams like it just take some time to show up.

OpenHAB-2.1.0.004-SNAPSHOT-DSM6-syno-noarch-0.001.spk
openHAB 2.1.0 Build #951
org.openhab.binding.robonect-2.1.0-SNAPSHOT.jar

sjka commented on addons/binding/org.openhab.binding.robonect/src/main/java/org/openhab/binding/robonect/RobonectClient.java:
```diff
@@ -0,0 +1,212 @@
+
+    private final Logger logger = LoggerFactory.getLogger(RobonectClient.class);
+
+    private HttpClient httpClient;
```
final

sjka commented on addons/binding/org.openhab.binding.robonect/src/main/java/org/openhab/binding/robonect/RobonectClient.java:
```diff
@@ -0,0 +1,212 @@
+
+    private HttpClient httpClient;
+
+    private ModelParser parser;
```
final

sjka commented on addons/binding/org.openhab.binding.robonect/src/test/java/org/openhab/binding/robonect/handler/RobonectHandlerTest.java:
```diff
@@ -0,0 +1,302 @@
+package org.openhab.binding.robonect.handler;
```
kudos for writing tests! 👍 
please move these into a separate test fragment though

reyem replied:
Usually makes my life easer and coding faster... ;-)
Will move it of course.

reyem commented on addons/binding/org.openhab.binding.robonect/ESH-INF/binding/binding.xml:
```diff
@@ -0,0 +1,11 @@
+
+	<name>Robonect Binding</name>
+	<description>This is the binding for Robonect.</description>
+	<author>marco.meyer</author>
```
will do.

reyem commented on addons/binding/org.openhab.binding.robonect/ESH-INF/i18n/robonect_xx_XX.properties:
```diff
@@ -0,0 +1,13 @@
+# FIXME: please substitute the xx_XX with a proper locale, ie. de_DE
```
makes sense

reyem commented on addons/binding/org.openhab.binding.robonect/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,254 @@
+        </channel-groups>
+        <config-description>
+            <parameter name="host" type="text" required="true">
+                <label>Host</label>
```
good point. did not know about the context :-)

reyem commented on addons/binding/org.openhab.binding.robonect/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,254 @@
+                <description>The user id if authentication has been enabled on the mower</description>
+            </parameter>
+            <parameter name="password" type="text" required="false">
+                <label>Password</label>
```
of course will add as well

reyem commented on addons/binding/org.openhab.binding.robonect/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,254 @@
+                <description>The interval for the binding to poll the mowers status information. Defaults to 30
+                    seconds.
+                </description>
+            </parameter>
```
cool, I was looking for that, but somehow I seem to be blind ;-) thx for mention

reyem commented on addons/binding/org.openhab.binding.robonect/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,254 @@
+        <item-type>Number</item-type>
+        <label>Mow duration</label>
+        <description>The mow duration</description>
+        <state readOnly="true"/>
```
ok.

reyem commented on addons/binding/org.openhab.binding.robonect/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,254 @@
+        <item-type>Number</item-type>
+        <label>Mow hours</label>
+        <description>The mow hours</description>
+        <state readOnly="true"/>
```
do as well

reyem commented on addons/binding/org.openhab.binding.robonect/README.md:
```diff
@@ -0,0 +1,132 @@
+Robonect is a piece of hardware which has to be put into your Husqvarna, Gardena and other branded automower and makes 
+it accessable in your internal network. More details about the Robonect module can be found at [robonect.de](http://www.robonect.de)
+
+This binding integrates mowers having the robonect module installed as thing into the openhab installation, allowing to
```
ok. So far I plan to contribute it to the openhab addons.... however, who knows :-)

reyem commented on addons/binding/org.openhab.binding.robonect/pom.xml:
```diff
@@ -0,0 +1,34 @@
+    <name>Robonect Binding</name>
+    <packaging>eclipse-plugin</packaging>
+
+    <dependencies>
```
another point I could not find when looking for it. Thank you for the link! Will change this of course

reyem replied:
I'm almost done locally with this refactoring for the tests... However, I stuck with hamcrest...

If I use in MANIFEST.MF
`Require-Bundle: org.hamcrest`

everything works ok but PMD/Checkstyle complains with

` ERROR found by checkstyle: .binding.robonect.test/META-INF/MANIFEST.MF:21 The MANIFEST.MF file must not contain any Require-Bundle entries. Instead, Import-Package must be used.`

If I remove the `Require-Bundle` part just leaving the following `Import-Package` declaration:

`
Import-Package: org.openhab.binding.robonect,
 org.openhab.binding.robonect.handler,
 org.openhab.binding.robonect.model,
 org.eclipse.smarthome.core.common.registry,
 org.eclipse.smarthome.core.events,
 org.eclipse.smarthome.core.thing.util,
 org.eclipse.smarthome.test.java,
 org.eclipse.smarthome.test.storage,
 org.mockito,
 org.mockito.invocation,
 org.mockito.stubbing,
 org.mockito.verification,
 org.hamcrest;core=split,
 org.junit;version="4.0.0",
 org.junit.runner,
 org.junit.runners,
 org.osgi.framework,
 org.osgi.service.device,
 org.slf4j,
 org.eclipse.smarthome.core.thing,
 org.eclipse.smarthome.core.types,
 org.eclipse.smarthome.core.library.types,
 org.eclipse.jetty.client,
 org.eclipse.jetty.http
`

The tests are not executed via maven anymore resulting in 

```
!ENTRY org.mockito 4 0 2017-06-29 09:17:00.889
!MESSAGE FrameworkEvent ERROR
!STACK 0
org.osgi.framework.BundleException: Could not resolve module: org.mockito [34]
  Unresolved requirement: Import-Package: COM.jrockit.reflect; resolution:="optional"
  Unresolved requirement: Import-Package: jrockit.vm; resolution:="optional"
  Unresolved requirement: Import-Package: org.hamcrest; version="[1.0.0,2.0.0)"

	at org.eclipse.osgi.container.Module.start(Module.java:434)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.incStartLevel(ModuleContainer.java:1582)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.incStartLevel(ModuleContainer.java:1561)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.doContainerStartLevel(ModuleContainer.java:1533)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.dispatchEvent(ModuleContainer.java:1476)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.dispatchEvent(ModuleContainer.java:1)
	at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:230)
	at org.eclipse.osgi.framework.eventmgr.EventManager$EventThread.run(EventManager.java:340)

!ENTRY org.openhab.binding.robonect.test 2 0 2017-06-29 09:17:01.079
!MESSAGE Could not resolve module: org.openhab.binding.robonect.test [37]
  Unresolved requirement: Import-Package: org.mockito
    -> Export-Package: org.mockito; bundle-symbolic-name="org.mockito"; bundle-version="1.9.5.v201311280930"; version="1.9.5"; uses:="org.mockito.stubbing,  org.hamcrest,  org.mockito.verification,  org.mockito.listeners,  org.mockito.exceptions.base,  org.mockito.exceptions,  org.mockito.configuration,  org.mockito.invocation"
       org.mockito [34]
         Unresolved requirement: Import-Package: COM.jrockit.reflect; resolution:="optional"
         Unresolved requirement: Import-Package: jrockit.vm; resolution:="optional"
         Unresolved requirement: Import-Package: org.hamcrest; version="[1.0.0,2.0.0)"


!ENTRY org.mockito 2 0 2017-06-29 09:17:01.079
!MESSAGE Could not resolve module: org.mockito [34]
  Unresolved requirement: Import-Package: COM.jrockit.reflect; resolution:="optional"
  Unresolved requirement: Import-Package: jrockit.vm; resolution:="optional"
  Unresolved requirement: Import-Package: org.hamcrest; version="[1.0.0,2.0.0)"
```

Any suggestions or samples on how to use moskito/hamcrest via pure imports? 
Also looking at the other bindings I could not find any not having at least hamcrest defined via Require-Bundle. So how can they skip this check style rule?

reyem commented on addons/binding/org.openhab.binding.robonect/src/main/java/org/openhab/binding/robonect/RobonectEndpoint.java:
```diff
@@ -0,0 +1,37 @@
+
+public class RobonectEndpoint {
+    
+    private String ipAddress;
```
completely agree

reyem commented on addons/binding/org.openhab.binding.robonect/src/main/java/org/openhab/binding/robonect/handler/RobonectHandler.java:
```diff
@@ -0,0 +1,401 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.openhab.binding.robonect.RobonectBindingConstants.CHANNEL_ERROR_CODE;
```
I use IntelliJ ):-)  ... So probably it will do when I finally will reformat everything using eclipse.

reyem commented on addons/binding/org.openhab.binding.robonect/src/main/java/org/openhab/binding/robonect/handler/RobonectHandler.java:
yep, agree...

reyem commented on addons/binding/org.openhab.binding.robonect/src/main/java/org/openhab/binding/robonect/handler/RobonectHandler.java:
well, that channel was thought to handle the case where the mower went offline and stays offline for a longer period. Unfortunately openhab/esh does not give access to the thing state in rules yet (there is an open issue about this https://github.com/eclipse/smarthome/issues/1654 ).

However, I will replace this with a trigger channel and a configuration for the timeout on how long the mower is expected to be offline.

reyem commented on addons/binding/org.openhab.binding.robonect/src/main/java/org/openhab/binding/robonect/handler/RobonectHandler.java:
Indeed...

reyem commented on addons/binding/org.openhab.binding.robonect/src/main/java/org/openhab/binding/robonect/handler/RobonectHandler.java:
good catch!

reyem commented on addons/binding/org.openhab.binding.robonect/src/main/java/org/openhab/binding/robonect/handler/RobonectHandler.java:
would agree

reyem commented on addons/binding/org.openhab.binding.robonect/src/main/java/org/openhab/binding/robonect/handler/RobonectHandler.java:
will be replaced

reyem commented on addons/binding/org.openhab.binding.robonect/src/main/java/org/openhab/binding/robonect/handler/RobonectHandler.java:
true

reyem commented on addons/binding/org.openhab.binding.robonect/src/main/java/org/openhab/binding/robonect/handler/RobonectHandler.java:
If you say so... But why is the ThingStatus defining this value than? If an exception would happen here, indeed the thing would not be initialized properly... So Offline might not be the right state.

reyem commented on addons/binding/org.openhab.binding.robonect/src/main/java/org/openhab/binding/robonect/model/MowerStatus.java:
```diff
@@ -0,0 +1,66 @@
+    UNKNOWN_15(15),
+    OFF(16),
+    SLEEPING(17),
+    OFFLINE(98),
```
as mentioned before, because of missing access to the things status in rules. however, will be replaced by trigger channel.

reyem commented on addons/binding/org.openhab.binding.robonect/src/main/java/org/openhab/binding/robonect/model/ModelParser.java:
```diff
@@ -0,0 +1,27 @@
+public class ModelParser {
+    
+    
+    private Gson gson;
```
why not.. :-)

reyem commented:
@sjka Thank you a lot for the valuable early review. I had quite some learnings and hope this will finally bring me closer to an accepted PR. Let me know once your module is installed and you tried it out, if you lack any features.

sjka commented on addons/binding/org.openhab.binding.robonect/src/main/java/org/openhab/binding/robonect/handler/RobonectHandler.java:
It's not 100% intuitive, but there are ThingStatus value which are set by the framework only, and some by the binding (i.e. ThingHandler) only. See https://www.eclipse.org/smarthome/documentation/concepts/things.html#thing-status

For unexpected runtime exceptions (i.e. programming bugs), there indeed is the status UNINITIALIZED/HANDLER_INITIALIZING_ERROR which I understand is pretty much what you would expect. Just let them "fall through" to the caller, the ThingManager will take care to set this status. 

On the other hand, there are several anticipated errors - and the more or less all wall into the category OFFLINE/CONFIGURATION_ERROR or OFFLINE/COMMUNICATION_ERROR, so the handler was working correctly but simply cannot talk to the device. Which in the sense of the thing handler still means: properly initialized. Just not usable. Therefore OFFLINE/...

sjka commented on addons/binding/org.openhab.binding.robonect/src/main/java/org/openhab/binding/robonect/handler/RobonectHandler.java:
I'm not so sure this is not possible - afaics #3001 added exactly this feature (see https://github.com/eclipse/smarthome/pull/3001#issuecomment-309250577). I have no idea though why the issue you mentioned is still open.

In any way I would prefer fixing the framework if this is not working sufficiently instead of defining channels in bindings to workaround this 😄

sjka commented:
Argh, your the merge of the latest master into your branch was not good. Now your PR became unreadable and it looks like it grew to 27k lines of code. Can your try rebasing your branch instead (i.e. `git rebase origin/master`)? 
It takes a few times to get your head around it, but in fact it is much easier and you will only have to solve merge conflicts once if you run into them (it simply stops there, you will get the chance to fix them and ` git add ...` them and continue with `git rebase --continue`. And if you feel you are really stuck in the middle or messed it up, a `git rebase --abort` will bring you back from where you started.

reyem commented:
ups, @sjka thank you for the tips. I tried to do the rebase. Then I had for some strange reason to pull again before I was able to push. I hope this is fine now? Or do I have to do some other actions? (I'm a git noob... :-/ )

martinvw commented:
@reyem Please note that, because it's marked [WIP], I do and did not review this PR completely yet.

If it's finished and ready to review please update the title and mention me so I can remove the "Work in progress" label.

reyem commented:
@martinvw, thank you for your notice and time. I did my final changes and the PR is ready for review. I think @sjka brought me already on the right path with an early review (thank you!), and some users are already using the binding in production like @DanielMalmgren and of course myself. So from the runtime it looks stable.

BOFH90 commented:
Hi, This Binding is great. I installed it yesterday morning and it worked perfectly. Today, my mower flaps between offline (just in openhab2) and online. But i can reach the Webinterface perfectly.
Error Message is:
2017-09-29 06:52:33.362 [ERROR] [ing.robonect.handler.RobonectHandler] - Unexpected exception. Setting thing offline
com.google.gson.JsonSyntaxException: java.io.EOFException: End of input at line 1 column 2

DanielMalmgren commented:
@BOFH90 Do you get any more relevant info in the log if you set the binding to bebug logging? Run the following in Karaf console:
log:set debug org.openhab.binding.robonect

BOFH90 commented:
@DanielMalmgren 
The error went away after restarting openHab2. If the error re-appears i can send you the Debug-Log.
The binding still worked at the time (communication and updated items between robonect and the item).

DanielMalmgren commented:
Too bad this binding didn't get it into 2.2 :-( Is it ok to give @kaikreuzer and @martinvw a little nudge?

martinvw commented on addons/binding/org.openhab.binding.robonect/src/main/java/org/openhab/binding/robonect/RobonectEndpoint.java:
```diff
@@ -0,0 +1,50 @@
+    
+    private final String password;
+    
+    private boolean useAuthentication = false;
```
And you always assign it so you can leave out the `false` here to make it final.

martinvw commented on addons/binding/org.openhab.binding.robonect/src/main/java/org/openhab/binding/robonect/handler/RobonectHandler.java:
```diff
@@ -0,0 +1,460 @@
+            } else {
+                switch (channelUID.getId()) {
+                    case CHANNEL_MOWER_NAME:
+                        if (command instanceof StringType) {
```
I think that nesting a if inside a switch inside an if inside a try is not nice, it could be improved by extracting both the handleRefreshCommand and the handleNormalCommands (or something like that) to their own methods. You can then leave the handling of the `RoboconnectException` and the if-else in the `handleCommand`

reyem replied:
ok, agree. done

martinvw commented on addons/binding/org.openhab.binding.robonect/src/main/java/org/openhab/binding/robonect/handler/RobonectHandler.java:
Please reformat

martinvw commented on addons/binding/org.openhab.binding.robonect/src/main/java/org/openhab/binding/robonect/handler/RobonectHandler.java:
Please reformat

reyem replied:
ok

martinvw commented on addons/binding/org.openhab.binding.robonect/src/main/java/org/openhab/binding/robonect/handler/RobonectHandler.java:
Do you really need/want internal state inside the anonymous inner class runnable?

reyem replied:
ok... refactored from anonymous to inner class..

martinvw commented on addons/binding/org.openhab.binding.robonect/src/main/java/org/openhab/binding/robonect/handler/RobonectHandler.java:
Can you import the `com.google.gson.JsonSyntaxException`

reyem replied:
of course :-)

martinvw commented on addons/binding/org.openhab.binding.robonect/src/main/java/org/openhab/binding/robonect/handler/RobonectHandler.java:
Please use [scheduleWithFixedDelay](http://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ScheduledExecutorService.html#scheduleWithFixedDelay(java.lang.Runnable,%20long,%20long,%20java.util.concurrent.TimeUnit)) which is preferred over scheduleAtFixedRate, see [Coding Guidelines D.3](http://docs.openhab.org/developers/development/guidelines.html#d-runtime-behavior)

reyem replied:
yep, done... Just wonder why I haven't seen this before

martinvw commented on addons/binding/org.openhab.binding.robonect/src/main/java/org/openhab/binding/robonect/handler/RobonectHandler.java:
We are disposing any way imho this could be a debug as well. `error`'s should be used sporadicaly, please check your usage of error:

In [E.4 of the coding guidelines](http://docs.openhab.org/developers/development/guidelines.html#e-logging) we listed what kind of logging level should be used when. Error should only be used:

> `error` logging should only be used to inform the user that something is tremendously wrong in his setup, the system cannot function normally anymore, and there is a need for immediate action.

reyem replied:
ok... changed to debug

cribskip commented:
Hi & thank you for a binding for robonect,

is there a place to download and test the current jar?

Thanks,
Sascha

reyem commented:
@cribskip 
This should work: https://openhab.jfrog.io/openhab/libs-pullrequest-local/org/openhab/binding/org.openhab.binding.robonect/2.2.0-SNAPSHOT/org.openhab.binding.robonect-2.2.0-SNAPSHOT.jar

reyem commented on addons/binding/org.openhab.binding.robonect.test/src/test/java/org/openhab/binding/robonect/RobonectClientTest.java:
```diff
@@ -0,0 +1,205 @@
+        when(responseMock.getContentAsString()).thenReturn(
+                "{\"successful\": true, \"name\": \"Mein Automower\", \"status\": {\"status\": 17, \"stopped\": false, \"duration\": 4359, \"mode\": 0, \"battery\": 100, \"hours\": 29}, \"timer\": {\"status\": 2, \"next\": {\"date\": \"01.05.2017\", \"time\": \"19:00:00\", \"unix\": 1493665200}}, \"wlan\": {\"signal\": -76}}");
+        client.getMowerInfo();
+        verify(httpClientMock, times(1)).newRequest("http://123.456.789.123/json?cmd=status");
```
The aim of this test is really to make sure the expected URL is called by the client for each command. Although I could verify the response, this makes limited sense, as the response is mocked. The parsing of the responses is tested in the ModelParserTest. So I do not want to duplicate here.

reyem commented on addons/binding/org.openhab.binding.robonect.test/.project:
```diff
@@ -0,0 +1,34 @@
+		</buildCommand>
+	</buildSpec>
+	<natures>
+		<nature>org.eclipse.m2e.core.maven2Nature</nature>
```
fixed... no clue how that got there :-)

reyem commented on addons/binding/org.openhab.binding.robonect.test/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,33 @@
+Bundle-Name:  Robonect Binding Tests
+Bundle-SymbolicName: org.openhab.binding.robonect.test;singleton:=true
+Bundle-Vendor: openHAB
+Bundle-Version: 2.2.0.qualifier
```
done.

reyem commented on addons/binding/org.openhab.binding.robonect.test/pom.xml:
```diff
@@ -0,0 +1,126 @@
+
+  <name>Robonect Binding Tests</name>
+
+  <properties>
```
removed.

reyem commented on addons/binding/org.openhab.binding.robonect.test/src/test/java/org/openhab/binding/robonect/RobonectClientTest.java:
```diff
@@ -0,0 +1,205 @@
+/**
+ * Copyright (c) 2010-2017 by the respective copyright holders.
```
thank you for the hint with the maven plugin! done

reyem commented on addons/binding/org.openhab.binding.robonect.test/src/test/java/org/openhab/binding/robonect/RobonectClientTest.java:
```diff
@@ -0,0 +1,205 @@
+import static org.mockito.Mockito.when;
+
+/**
+ * @author Marco Meyer - Initial contribution
```
ok.. was a bit lazy with the test classes :-|
Done

reyem commented on addons/binding/org.openhab.binding.robonect.test/src/test/java/org/openhab/binding/robonect/handler/RobonectHandlerTest.java:
```diff
@@ -0,0 +1,321 @@
+ */
+public class RobonectHandlerTest {
+
+    @Mock
```
ups. for sure not!
removed.

reyem commented on addons/binding/org.openhab.binding.robonect.test/src/test/java/org/openhab/binding/robonect/handler/RobonectHandlerTest.java:
```diff
@@ -0,0 +1,321 @@
+public class RobonectHandlerTest {
+
+    @Mock
+    private RobonectHandler testObj;
```
I like subject. Renamed it in all 3 test classes.

reyem commented on addons/binding/org.openhab.binding.robonect/.project:
```diff
@@ -0,0 +1,39 @@
+		</buildCommand>
+	</buildSpec>
+	<natures>
+		<nature>org.eclipse.m2e.core.maven2Nature</nature>
```
removed... also the build command above..

reyem commented on addons/binding/org.openhab.binding.robonect/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,29 @@
+Bundle-Name: Robonect Binding
+Bundle-SymbolicName: org.openhab.binding.robonect;singleton:=true
+Bundle-Vendor: openHAB
+Bundle-Version: 2.2.0.qualifier
```
no prob. done

reyem commented on addons/binding/org.openhab.binding.robonect/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,29 @@
+ org.apache.commons.lang
+Service-Component: OSGI-INF/*.xml
+Export-Package: org.openhab.binding.robonect,
+ org.openhab.binding.robonect.handler, org.openhab.binding.robonect.model
```
nope. I had quite some troubles with getting the classes available in the test module via OSGI. So that was a left over. removed.

reyem commented on addons/binding/org.openhab.binding.robonect/OSGI-INF/RobonectHandlerFactory.xml:
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
OK. Have to learn about OSGI annotations first. If you have a quick link on where to look up, this would be very welcome.

reyem replied:
OK, found out... :-)
Easy. Done.

reyem commented:
😕 messed it up again... will fix it another time...

reyem commented on addons/binding/pom.xml:
```diff
@@ -72,6 +72,8 @@
     <module>org.openhab.binding.rme</module>
     <module>org.openhab.binding.rfxcom</module>
     <module>org.openhab.binding.rfxcom.test</module>
+    <module>org.openhab.binding.robonect</module>
```
ok, done.

reyem commented on addons/binding/org.openhab.binding.robonect/src/main/java/org/openhab/binding/robonect/handler/RobonectHandler.java:
```diff
@@ -0,0 +1,460 @@
+
+    private final Logger logger = LoggerFactory.getLogger(RobonectHandler.class);
+
+    public RobonectHandler(Thing thing) {
```
done

reyem commented on addons/binding/org.openhab.binding.robonect/src/main/java/org/openhab/binding/robonect/handler/RobonectHandler.java:
```diff
@@ -0,0 +1,460 @@
+                }
+
+            }
+            updateStatus(ThingStatus.ONLINE);
```
well, there are cases where the mower exits wifi range during an update leading to an exception, setting the thing offline. No, if the command is sent again, and it was successful, the mower would be online again. But of course it also is updated by the polling thread. Maybe could maybe removed here. But does it any harm? Because the polling thread might just poll every 30s. So to wait 30s in case the mower just exited range shortly is also bad, if the command is retired immediately and also successful..

reyem commented on addons/binding/org.openhab.binding.robonect/src/main/java/org/openhab/binding/robonect/handler/RobonectHandler.java:
That is a leftover, when there was some effects in production which I was first not able to follow up as there was no log. So I was, guessing an exception was thrown but not matched anywhere. The question here is, what would happen. Does open hab have a general exception handler which would notify about such cases? Also I would have to catch at least InteruptedExceptions as they are potentially thrown by the RobonectClient

reyem commented on addons/binding/org.openhab.binding.robonect/src/main/java/org/openhab/binding/robonect/model/cmd/StopCommand.java:
```diff
@@ -0,0 +1,28 @@
+    public String toCommandURL(String baseURL) {
+        return baseURL + "?cmd=stop";
+    }
+}
```
🆗 done

reyem commented on addons/binding/org.openhab.binding.robonect/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,297 @@
+                          xmlns:thing="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0"
+                          xsi:schemaLocation="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0 http://eclipse.org/smarthome/schemas/thing-description-1.0.0.xsd">
+    
+    <thing-type id="mower">
```
ok, done... although for some strange reason checkstyle still complains about line 3... No clue where it finds non-tab space for structuring...

reyem commented on addons/binding/org.openhab.binding.robonect/README.md:
```diff
@@ -0,0 +1,119 @@
+| `mowerStatus#battery`    | Number                    | Retrieves the current battery status in percent                                    |
+| `mowerStatus#duration`   | Number                    | Retrieves the duration of the current status (see `mowerStatus#status`) of the mower |
+| `mowerStatus#hours`      | Number                    | Retrieves the number of hours of mowing operation                                  |
+| `mowerStatus#mode`       | String                    | Retrieves or  sets the mode of the mower. Possible values retrieval values are <ul><li>HOME</li><li>AUTO</li><li>MAN</li></ul> In addition he channel allows to set following values for triggering special actions <ul><li>EOD : triggers the "end of day" mode. The mower will switch in to the HOME mode and stay int this mode for the rest of the ...
```
I would like, if you can tell me how. I was following the advices from here:
https://stackoverflow.com/questions/19950648/how-to-write-lists-inside-a-markdown-table

Alternatively I could remove the list. But I think it doesn't improve readability. And as github and most IDE anyway already have markdown formatted live previews, I beliefe it is better to have html in markdown than removing the lists..

Confectrician replied:
The docs repo (or to be more specific jekyll) can handle that too,
alltough we wanted to get away from html to complete markdown some time.

I haven't found a really usabel situation for this myself yet and we have some more binding cases where we have html in markdown tables.

reyem commented on addons/binding/org.openhab.binding.robonect/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,297 @@
+        </config-description>
+    </thing-type>
+
+    <channel-group-type id="mowerInfoType">
```
ok... I'm happy for proposals for alternatives. So far my idea was to use the groups for structuring the channels to logical units.

martinvw replied:
@kaikreuzer can you please validate / check this usage of channel groups, given the time I spent on other things somehow my mind still does not have a matured definition of the channel groups 😉

reyem commented:
@martinvw thank you for your review. I think I fixed all the points, except
* channel-groups: Don't know how you want me to change them?
* Readme.md: HTML List in markdown table. Do not know a better solution.

Also I fixed all check style issues I could. Sorry for not doing this before. However, for some reason on the XML Files check style still complains about spaces I cannot find :-|

How shall we proceed?

martinvw commented:
> Readme.md: HTML List in markdown table. Do not know a better solution.

Could I get some help from one of the README guroes :-) @Confectrician @ThomDietrich

Confectrician commented:
List in a table is possibly really difficult and additionally hard to maintain afterwards too.
I will take a closer look at the contents tommorow.

Maybe we can find a better structure, but i need to get some content knowledge before able to tell something about it.

Confectrician commented:
After taking a closer look i would say that keeping the html is the easiest Solution for now.
There are too many table rows affected to cut those parts out of the table and move them somewhere else as explanation.

Confectrician commented on addons/binding/org.openhab.binding.robonect/README.md:
````diff
@@ -0,0 +1,123 @@
+Items file `.items`
+
+```
+String mowerName "Mower name" {robonect:mower:automower:mowerInfo#name"}
````
Maybe you could format the items definitions a bit to improve the readability.

reyem replied:
good idea. done.

Confectrician commented on addons/binding/org.openhab.binding.robonect/README.md:
````diff
@@ -0,0 +1,123 @@
+
+Map transformation for mower status (`robonect_status.map`)
+
+```
````
Please use a specific codefence here.

` ```text `

reyem replied:
done

reyem replied:
done

Confectrician commented on addons/binding/org.openhab.binding.robonect/README.md:
````diff
@@ -0,0 +1,123 @@
+
+Items file `.items`
+
+```
````
Please use a specific codefence here.

` ```java `

reyem replied:
done

Confectrician commented on addons/binding/org.openhab.binding.robonect/README.md:
````diff
@@ -0,0 +1,123 @@
+
+Things file `.things`
+
+```
````
Please use a specific codefence here.

` ```java `

reyem replied:
done

Confectrician commented on addons/binding/org.openhab.binding.robonect/README.md:
````diff
@@ -0,0 +1,123 @@
+
+An example things configuration might look like
+
+```
````
Please use a specific codefence here.

` ```java `

reyem replied:
added... However, why java? Shouldn't this be 'text' as well

reyem replied:
done

reyem commented:
@Confectrician Thank you for the review and the tipps! I think this improved the readability quite a bit. And the tablesgenerator.com is really helpful!

reyem commented:
@sjka Thank you for the additional feedback points. will look into them. I was not brave enough yet for the robonect 1.0 beta :-) Thus I did not recognize it changed. The JSON you posted is definitely failing due a structural change. Although that for itself would be easy to fix, I'm afraid other APIs changed as well. So I'm not sure if I manage to include compatibility with 1.0 beta still in this PR :-(

reyem commented on addons/binding/org.openhab.binding.robonect/src/main/java/org/openhab/binding/robonect/handler/RobonectHandler.java:
true, do a small refactoring not passing the config at all in the constructor.

reyem commented on addons/binding/org.openhab.binding.robonect/src/main/java/org/openhab/binding/robonect/handler/RobonectHandler.java:
Hmmm, interesting. there is mentioned "unless there are specific configuration requirements, the shared instance should be used". The binding modifies the authentication store to implement "preemptive" authentication (jetty was not supporting that out of the box in the jetty version included in ESH that time...) So would that be a specific configuration or not?

```java
private void addPreemptiveAuthentication(HttpClient httpClient, RobonectEndpoint endpoint) {
        AuthenticationStore auth = httpClient.getAuthenticationStore();
        URI uri = URI.create(baseUrl);
        auth.addAuthenticationResult(new BasicResult(HttpHeader.AUTHORIZATION, uri, "Basic " + B64Code
                .encode(endpoint.getUser() + ":" + endpoint.getPassword(), StandardCharsets.ISO_8859_1)));
    }
```

So shall I really do this, or is there a risk of breaking the shared clients...

sjka replied:
Good point, I missed that one. No, I agree, it doesn't sound like a good idea anymore...

reyem commented on addons/binding/org.openhab.binding.robonect/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,289 @@
+			<parameter name="pollInterval" type="integer" required="false" unit="s">
+				<label>Polling interval</label>
+				<default>30</default>
+				<description>The interval for the binding to poll the mowers status information.
```
Actually I think I can leave away the "seconds" as the unit is specified. Am I right?

sjka replied:
👍

reyem commented on addons/binding/org.openhab.binding.robonect/src/main/java/org/openhab/binding/robonect/handler/RobonectHandler.java:
agree.

reyem commented on addons/binding/org.openhab.binding.robonect/src/main/java/org/openhab/binding/robonect/handler/RobonectHandler.java:
yep

reyem commented on addons/binding/org.openhab.binding.robonect/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,289 @@
+		<item-type>Number</item-type>
+		<label>Humidity</label>
+		<description>The humidity in the mower</description>
+		<state readOnly="true" pattern="%d '%'"/>
```
ok

reyem commented on addons/binding/org.openhab.binding.robonect/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,289 @@
+	<thing-type id="mower">
+		<label>Mower</label>
+		<description>Mower robot connected via robonect module</description>
+		<channel-groups>
```
That is an interesting point, as @martinvw is not sure if it is even an abuse of groups. I'm confused. I like structuring information. However, I don't want to abuse the groups. Therefore now I'm really curious if @kaikreuzer could give the link to the conceptual idea of groups so we can solve this topic soon.

kaikreuzer replied:
I agree with @sjka that this looks much too fine-grained. Originally, channel-groups were introduced in order to reuse sets of channels and not for categorising the existing list. It is also ok to use groups if a device consists out of rather independent sub-devices, but I would refrain from the case here. Maybe you could at least reduce it to two groups like "Maintainance" and "Operation" if that makes sense.

reyem commented on addons/binding/org.openhab.binding.robonect/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,289 @@
+			<channel id="mode" typeId="modeType"/>
+			<channel id="status" typeId="mowerStatusType"/>
+			<channel id="started" typeId="startedType"/>
+			<channel id="offlineTrigger" typeId="offlineTriggerType"/>
```
Well, the idea here is, that the binding provides functionality of allowing certain duration of "offline" and trigger an alert if this is tine exceeded. The real world use case is, that in many cases the mower actually looses the signal to Wifi as it exits the range but finally will come back. If not, most probably the mower sticks somewhere. Now the "community" at robonect.de wished the functionality in the binding to not have to code custom rules for this condition. I'm happy to remove it, if you provide a sample rule which handles this use case properly. Then I can put this in the Readme.md...

sjka replied:
Okay, I see. 

With the new rule engine it would be simple as the binding could bring a `TriggerModule` which encapsulates exactly this functionality. But as most of us are still stuck on the old one, I agree it would be more convenient to keep this channel.

reyem commented on addons/binding/org.openhab.binding.robonect/README.md:
```diff
@@ -0,0 +1,117 @@
+|-------------------------|------------------|---------------------------|
+| Husqvarna Automower 105 | Robonect Hx      | 0.9c, 0.9e                |
+| Husqvarna Automower 315 | Robonect Hx      | 0.9e, 1.0 preview         |
+
```
OK, will try to fix this ...

sjka commented on addons/binding/org.openhab.binding.robonect/src/main/java/org/openhab/binding/robonect/handler/RobonectHandler.java:
At least that was my naive assumption. What else should interrupt the threads?

reyem replied:
ok, refactored to catch in the RobonectClient directly and throw a RobonectCommunicationException in this case. Simplified it a lot.

sjka commented:
> I was not brave enough yet for the robonect 1.0 beta

Yeah, the more I think about [it](http://www.robonect.de/viewtopic.php?f=5&t=1291), the less I understand why I did that... 😩

> So I'm not sure if I manage to include compatibility with 1.0 beta still in this PR 

Sure. It makes sense to first finish 0.9e support and get that merged. You may still add the 420 with `0.9e, 1.0 preview` to the list right now as that worked pretty well last year.

If I can help you somehow, e.g. by sending you all the JSON responses and/or testing, I'm happy to help as I so far see no pressure to downgrade again. Also in a separate, follow-up PR.

reyem commented:
@sjka : I fixed some of your review points. But most interesting for you is that I fixed the NPE. Now the version and serial properties will be set to 'n/a'. I think this is better than an NPE. Would be cool if you could test the basic channels in the mean time so we get an idea of what the remaining effort on the next PR would be ;-)

thokl71 commented:
Hello, I tried the binding with my Automower 105. Sending home or start mowing works great. My problem is that the mower is beeping all the time so at night I have to switch it off. I created a rule that notifys me on every status change. When set to not mow I get 1 and 17 every minute, one time 1 a minute later 17. And every minute the mower beeps. I think it has something to do with the polling. Anyone else with this problem?

martinvw commented:
Hello what is the status in you opinion, are you waiting for me? Are you gong to pick up @sjka remaining comments? Please let me know, thanks for you hard work.

reyem commented on addons/binding/org.openhab.binding.robonect/src/main/java/org/openhab/binding/robonect/handler/RobonectHandler.java:
ok, now just wrap the start() and refreshVersionInfo() methods.

reyem replied:
after refactoring the InterruptedException, just start() is in the try-catch block :-)

sjka replied:
Hmmm, looks like this was going a little too far: I started the binding at the exact point in time when my mover was outside of the wifi coverage. The thing stayed in UNINTIALIZED because of this exception:

```
2018-05-02 13:21:34.740 [ERROR] [nal.common.AbstractInvocationHandler] - An error occurred while calling method 'ThingHandler.initialize()' on 'org.openhab.binding.robonect.handler.RobonectHandler@519b4a71': Could not send command http://shawn.gdhsn.kaufleute.name/json?cmd=version
org.openhab.binding.robonect.internal.RobonectCommunicationException: Could not send command http://shawn.gdhsn.kaufleute.name/json?cmd=version
	at org.openhab.binding.robonect.internal.RobonectClient.sendCommand(RobonectClient.java:308)
	at org.openhab.binding.robonect.internal.RobonectClient.getVersionInfo(RobonectClient.java:317)
	at org.openhab.binding.robonect.handler.RobonectHandler.refreshVersionInfo(RobonectHandler.java:339)
	at org.openhab.binding.robonect.handler.RobonectHandler.initialize(RobonectHandler.java:392)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)[:1.8.0_152]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)[:1.8.0_152]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)[:1.8.0_152]
	at java.lang.reflect.Method.invoke(Method.java:498)[:1.8.0_152]
	at org.eclipse.smarthome.core.internal.common.AbstractInvocationHandler.invokeDirect(AbstractInvocationHandler.java:153)[49:org.eclipse.smarthome.core:0.10.0.201804301608]
	at org.eclipse.smarthome.core.internal.common.Invocation.call(Invocation.java:53)[49:org.eclipse.smarthome.core:0.10.0.201804301608]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)[:1.8.0_152]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)[:1.8.0_152]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)[:1.8.0_152]
	at java.lang.Thread.run(Thread.java:748)[:1.8.0_152]
Caused by: java.util.concurrent.ExecutionException: java.net.SocketTimeoutException: Connect Timeout
	at org.eclipse.jetty.client.util.FutureResponseListener.getResult(FutureResponseListener.java:118)
	at org.eclipse.jetty.client.util.FutureResponseListener.get(FutureResponseListener.java:110)
	at org.eclipse.jetty.client.HttpRequest.send(HttpRequest.java:676)
	at org.openhab.binding.robonect.internal.RobonectClient.sendCommand(RobonectClient.java:291)
```

The exception dropped trough to the framework. As a SocketTimeoutException (or even: IOException) is something expected, it should be rather caught and the thing set to OFFLINE/COMMUNICATION_ERROR in such case.

reyem replied:
ups, lucky me, that you tested this in the right moment :-)
I now moved the loadVersionInfo again in the polling thread, and reload it each time the thing went offline. That way it would also pick up new properties if you do a firmware upgrade (or more probably downgrade at the moment ;-) ), without the need to restart the binding.

reyem commented on addons/binding/org.openhab.binding.robonect/src/main/java/org/openhab/binding/robonect/handler/RobonectHandler.java:
actually true, although I will leave this for now.

reyem commented on addons/binding/org.openhab.binding.robonect/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,289 @@
+	</channel-type>
+
+	<channel-type id="durationType">
+		<item-type>Number</item-type>
```
changed the description to "number of seconds ...." ) . Or is there something like a Time Item type? Can just find DateTime, but that does not match.

reyem commented on addons/binding/org.openhab.binding.robonect/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,289 @@
+	</channel-type>
+
+	<channel-type id="hoursType">
+		<item-type>Number</item-type>
```
!? same here?

reyem commented on addons/binding/org.openhab.binding.robonect/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,289 @@
+	</channel-type>
+
+	<channel-type id="temperatureType">
+		<item-type>Number</item-type>
```
OK, now I start understanding.... Just learned about the QuantityTypes in ESH... interesting. But could not find something like "Time" here:
https://www.eclipse.org/smarthome/documentation/concepts/units-of-measurement.html

reyem commented:
@martinvw 
I will fix most of the review points from @sjka as they make lots of sense. Probably I can do it this week. However, I still miss the feedback about the channel-groups. You mentioned @kaikreuzer shall have a look to judge if they are used in a wrong way. Other than that I think the binding is ready. So can you somehow trigger/accelerate the channel-group discussion?
thx

kaikreuzer commented on addons/binding/org.openhab.binding.robonect/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,289 @@
+		<channels>
+			<channel id="remoteStart" typeId="remoteStartType"/>
+			<channel id="afterMode" typeId="modeType"/>
+			<channel id="start" typeId="timeStringType"/>
```
Or even better do a "startTimeType" and end "endTimeType" as "timeStringType" as "timeStringType" seems to be too generic.
Btw, those types should also be of item type "DateTime" and not "String" - you can create them with hh:mm format (and without a date).

reyem replied:
OK, changed to DateTime...

reyem commented:
@thokl71 

> My problem is that the mower is beeping all the time so at night I have to switch it off. I created a rule that notifys me on every status change. When set to not mow I get 1 and 17 every minute, one time 1 a minute later 17. And every minute the mower beeps. I think it has something to do with the polling. Anyone else with this problem?

Although the binding polls the API, the beep is caused by the robonect module itself. I have a 105 myself running firmware 0.9e and do not have that issue. It just beeps if I change the mode from the binding. Maybe ask in the robonect forum(http://robonect.de) for advice on how to get rid of the beep...

reyem commented on addons/binding/org.openhab.binding.robonect/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,289 @@
+	</channel-type>
+
+	<channel-type id="humidityType">
+		<item-type>Number</item-type>
```
is dimensionless better than not specified/ Number without measurement type?

kaikreuzer replied:
Yes, it is Dimension=Dimensionless with Unit=Percent.
So you define the type as `Number:Dimensionless` and send state updates as `QuantityType` with unit `Percent`.

reyem replied:
👍 thanks. Done.

DanielMalmgren commented:
@reyem 
> Although the binding polls the API, the beep is caused by the robonect module itself. I have a 105 myself running firmware 0.9e and do not have that issue. It just beeps if I change the mode from the binding. Maybe ask in the robonect forum(http://robonect.de) for advice on how to get rid of the beep...

Yep, this seems to be a side effect of upgrading to the latest betas. Also noticed it yesterday (I mentioned it in the Robonect thread of the OH forum). I also threw the question out in the Robonect forum ([http://robonect.de/viewtopic.php?f=9&t=365&p=12265#p12265](url)) but haven't got any response yet. I've had to remove the AM from OH for now because of this, so I hope Fabian solves it...

reyem commented on addons/binding/org.openhab.binding.robonect/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,289 @@
+	<channel-type id="temperatureType">
+		<item-type>Number</item-type>
+		<label>Temperature</label>
+		<description>The tempurature in the mower</description>
```
done

reyem commented on addons/binding/org.openhab.binding.robonect/src/main/java/org/openhab/binding/robonect/handler/RobonectHandler.java:
👍 done

reyem commented on addons/binding/org.openhab.binding.robonect/src/main/java/org/openhab/binding/robonect/handler/RobonectHandler.java:
yep

reyem commented on addons/binding/org.openhab.binding.robonect/src/main/java/org/openhab/binding/robonect/handler/RobonectHandler.java:
👍 done

reyem commented on addons/binding/org.openhab.binding.robonect/src/main/java/org/openhab/binding/robonect/handler/RobonectHandler.java:
yep. done

reyem commented on addons/binding/org.openhab.binding.robonect/README.md:
```diff
@@ -0,0 +1,117 @@
+| `mowerStatus#offlineTrigger` | Trigger   | Is triggered if the mower is longer than the configured offlineTriggerTimeout offline. This might help if the mower is somewhere in error state but does not have a signal                                                                                                                                                                                         ...
+| `job#remoteStart`            | String    | Sets the remote start type for a JOB. Possible values are <ul><li>STANDARD : use the remote start option defined in the mower settings</li><li>REMOTE_1 : start from remote 1</li><li>REMOTE_2 : start from remote 2</li></ul>                                                                                                                                     ...
+| `job#afterMode`              | String    | Sets the mode the mower should be put in after a mowing Job. (See JOB on channel `owerStatus#mode`)                                                                                                                                                                                                                                                                ...
+| `job#start`                  | String    | Sets the start time for the next job in the form hh:mm                                                                                                                                                                                                                                                                                                             ...
```
indeed, I agree with this. I run into the issue already with another binding where we had to drop the "hack"... But this resulted in a missing feature on that binding now, which makes it unusable for me. So I stay on SNAPSHOT until this concept is ready and a proper PR can be made and is accepted. So over the thumb 2 years from now :-(
Thus I would like to keep the channels in this binding and remove as soon as we have a proper rule action concept and a PR for this.

reyem commented on addons/binding/org.openhab.binding.robonect/src/main/java/org/openhab/binding/robonect/handler/RobonectHandler.java:
yeah, removed this full catch block now. The InterruptedException form the httpClient.send(...) is now packed into a RobonectCommunicationException.

reyem commented:
@martinvw @kaikreuzer @sjka 
So, finally I fixed all the review points. I have to agree that the channel-groups were overkill. After refactoring, it looks somehow cleaner. Thank you for all your valuable input. The only thing I miss now, is a final test with a real mower. As my testing raspberry just broke down, this will need a moment. But I would be very happy if you could already do a final review, such that we can hopefully close this PR soon to provide an official binding for the mowing season :-) 

@all-snapshot-users: Please check the Readme.md for the new channel names. Unfortunately your item files will not work anymore with the newest SNAPSHOT. Sorry :-/

sjka commented:
I just updated to the latest snapshot and it's indeed a great improvement 👍 
Apart from the `SocketTimeoutException` which went uncaught during `initialize()` this bindings behaves pretty solid now. 
Thanks for the great work!

marco-assentis commented:
From my view, all requested changes are done. Thus I guess 👍 
@martinvw @kaikreuzer hope you can finally approve :-)

reyem commented:
from my side all requested changes are done... So @martinvw @kaikreuzer would be great if you find the time for the final review and approve. :-)

reyem commented on addons/binding/org.openhab.binding.robonect/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,256 @@
+			<channel id="timer-status" typeId="timerStatusType"/>
+			<channel id="timer-next" typeId="nextTimerType"/>
+
+			<channel id="job-remote-start" typeId="remoteStartType"/>
```
I like the switch proposal! Just have to find out on how to parameterize channels though. But I guess I will find it in the docu or other bindings...

kaikreuzer replied:
Yeah, the docu is still a bit weak on that, but check the [Astro binding](https://github.com/eclipse/smarthome/blob/master/extensions/binding/org.eclipse.smarthome.binding.astro/ESH-INF/thing/channels.xml#L91-L97) as an example with [this config for a channe](https://github.com/eclipse/smarthome/blob/master/extensions/binding/org.eclipse.smarthome.binding.astro/ESH-INF/config/config.xml#L20-L34).

reyem replied:
So, done as well... Nice would be if one could configure multiple channels of theses, but I think this is now a good intermediate way until the action concept is ready.

reyem commented on addons/binding/org.openhab.binding.robonect/src/main/java/org/openhab/binding/robonect/handler/RobonectHandler.java:
This is something @sjka already mentions. But after mentioning that I had to customize the authentication store of the HttpClient instance, he was also skeptical if this is gonna be a good idea.

Concrete I do the following in RobonectClient to get Preemptive Authentication working

```java
AuthenticationStore auth = httpClient.getAuthenticationStore();
        URI uri = URI.create(baseUrl);
        auth.addAuthenticationResult(new BasicResult(HttpHeader.AUTHORIZATION, uri, "Basic " + B64Code
                .encode(endpoint.getUser() + ":" + endpoint.getPassword(), StandardCharsets.ISO_8859_1)));
```

Do you think this will not have a side effect on others using this client? theoretically the authorizationResult is stored together with the URL, thus it could work..

During 2.2 development this was requrired, as the version of jetty client was quite old. Did jetty client version change since 2.2.0?

kaikreuzer replied:
No, Jetty is still 9.3.15(+) - so yes, I agree that it is probably safer to keep a separate client instance.
Nonetheless, you might want to retrieve it from `HttpClientFactory.createHttpClient()` as this will use a smaller thread pool (that can be configured for the system) and named threads, so that it is easier to find them in debugging sessions.

reyem replied:
ok, refactored to use factory

reyem commented on addons/binding/org.openhab.binding.robonect/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,256 @@
+
+	<channel-type id="versionType">
+		<item-type>String</item-type>
+		<label>Robonect version</label>
```
dito

reyem commented on addons/binding/org.openhab.binding.robonect/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,256 @@
+
+	<channel-type id="compiledType">
+		<item-type>String</item-type>
+		<label>Robonect version</label>
```
sure

reyem commented on addons/binding/org.openhab.binding.robonect/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,256 @@
+
+	<channel-type id="commentType">
+		<item-type>String</item-type>
+		<label>Robonect firmware comment</label>
```
as well

reyem commented on addons/binding/org.openhab.binding.robonect/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,256 @@
+	<channel-type id="nameType">
+		<item-type>String</item-type>
+		<label>Mower name</label>
+		<description>The name of the mower</description>
```
hmmm, here I'm unsure. Although the use case to change the name via OH is rather hypothetic. But if I want to get this via in the properties section I have to do an extra call. On the other hand handling it with the other channels via polling does not cost anything. The name is gettable and settable, thus, why not leave it as channel?

kaikreuzer replied:
Ok, right, as it is settable, it indeed makes sense as a channel!

reyem commented on addons/binding/org.openhab.binding.robonect/src/main/java/org/openhab/binding/robonect/handler/RobonectHandler.java:
it's the httpClient.start() which may throw an exception here... I guess with the shared httpClient instance this would disappear as well, right. Just still the question if PreEmptive Authentication counts as "specific configuration requirements" and therefore I cannot use it. In this case I understand I would throw the exception. Can I wrap it into a RuntimeException?

kaikreuzer replied:
Yeah, wrapping it in a RuntimeException should be fine in this case (as the Exception from Jetty is anyhow very unspecific).

DanielMalmgren replied:
I don't know if it's related here, but every time the binding tries to poll my mower and doesn't succeed I get an error in the log, "Fatal transport error: java.util.concurrent.TimeoutException". Not being able to reach it should be considered quite normal, classing it as an error doesn't seem right.

kaikreuzer replied:
Sounds like a good example why more finegrained exceptions should be caught and handled differently - a TimeoutException should indeed set the Thing to OFFLINE-COMMUNICATION_ERROR and do a periodic retry.

DanielMalmgren replied:
Not too sure about that @kaikreuzer . There is a setting for offline timeout at Thing level, described by @reyem as "The maximum time the mower may be offline before the offline trigger is triggered." The mower being temporarily outside of range is normal and it's not an error unless it's not still unreachable after this timeout.

reyem replied:
actually this will not happen anymore. This was just possible before, as I was calling refreshInfo already the first time here in a previous version in the binding. But the refactoring removed this. So that should be ok.. now

reyem commented on addons/binding/org.openhab.binding.robonect/README.md:
```diff
@@ -0,0 +1,118 @@
+| `mode`                       | String    | Retrieves or  sets the mode of the mower. Possible values retrieval values are <ul><li>HOME</li><li>AUTO</li><li>MAN</li></ul> In addition he channel allows to set following values for triggering special actions <ul><li>EOD : triggers the "end of day" mode. The mower will switch in to the HOME mode and stay int this mode for the rest of the day</li><li ...
+| `status`                     | Number    | Retrieves the current mower status which can be <ul><li>0 : DETECTING_STATUS</li><li>1 : PARKING</li><li>2 : MOWING</li><li>3 : SEARCH_CHARGING_STATION</li><li>4 : CHARGING</li><li>5 : SEARCHING</li><li>6 : UNKNOWN_6</li><li>7 : ERROR_STATUS</li><li>16 : OFF</li><li>17 : SLEEPING</li><li>98 : OFFLINE (Binding cannot connect to mower)</li><li>99 : UNKNO ...
+| `started`                    | Switch    | Switches the mower ON (analog to start button on mower) or OFF (analog to stop button on mower).                                                                                                                                                                                                                                                                   ...
+| `offlineTrigger`             | Trigger   | Is triggered if the mower is longer than the configured offlineTriggerTimeout offline. This might help if the mower is somewhere in error state but does not have a signal                                                                                                                                                                                         ...
```
you mean removing the offlineTrigger from the table and create a paragraph after the table describing it separately? Or leave it in the table but mention in addition after the table?

kaikreuzer replied:
Yes, removing it from the table and describe it separately.

reyem replied:
ok, done

kaikreuzer commented on addons/binding/org.openhab.binding.robonect/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,256 @@
+			<channel id="mowing-hours" typeId="hoursType"/>
+
+			<channel id="mode" typeId="modeType"/>
+			<channel id="started" typeId="startedType"/>
```
Is this about the "start" command (http://www.robonect.de/viewtopic.php?f=11&t=39)?
If so, I would simply suggest to name it "start" - this would be in line with the module API and also not appear to be read-only for the Thing.

reyem replied:
:-) Thank you. So simple but logic. Somehow I was always looking at the name of the status message and thus took the wrong base. So finally simply renamed it to "start" as you suggested.

reyem commented on addons/binding/org.openhab.binding.robonect/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,256 @@
+		<item-type>Number:Time</item-type>
+		<label>Status duration</label>
+		<description>The number of seconds the mower is in the current status.</description>
+		<state readOnly="true" pattern="%d seconds"/>
```
done

reyem commented on addons/binding/org.openhab.binding.robonect/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,256 @@
+		<item-type>Number:Temperature</item-type>
+		<label>Temperature</label>
+		<description>The temperature inside the mower</description>
+		<state readOnly="true" pattern="%d °C"/>
```
done

reyem commented on addons/binding/org.openhab.binding.robonect/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,256 @@
+		<description>The mower mode</description>
+		<state readOnly="false">
+			<options>
+				<option value="AUTO">AUTO</option>
```
done

reyem commented on addons/binding/org.openhab.binding.robonect/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,256 @@
+	<channel-type id="endTimeType">
+		<item-type>DateTime</item-type>
+		<label>The end time for a mowing job</label>
+		<description>The time to job should end</description>
```
done, resp. removed anyways

reyem commented on addons/binding/org.openhab.binding.robonect/README.md:
```diff
@@ -0,0 +1,118 @@
+# Robonect Binding
+
+Robonect is a piece of hardware which has to be put into your Husqvarna, Gardena and other branded automower and makes 
+it accessible in your internal network. More details about the Robonect module can be found at [robonect.de](http://www.robonect.de)
```
done

reyem commented on addons/binding/org.openhab.binding.robonect/README.md:
```diff
@@ -0,0 +1,118 @@
+
+following configuration settings are supported for the `mower` thing.
+
+| property name  | mandatory | description                                                                                       |
```
done

reyem commented on addons/binding/org.openhab.binding.robonect/README.md:
```diff
@@ -0,0 +1,118 @@
+| `battery`                    | Number    | Retrieves the current battery status in percent                                                                                                                                                                                                                                                                                                                    ...
+| `status-duration`            | Number    | Retrieves the duration of the current status (see `status`) of the mower                                                                                                                                                                                                                                                                                           ...
+| `mowing-hours`               | Number    | Retrieves the number of hours of mowing operation                                                                                                                                                                                                                                                                                                                  ...
+| `mode`                       | String    | Retrieves or  sets the mode of the mower. Possible values retrieval values are <ul><li>HOME</li><li>AUTO</li><li>MAN</li></ul> In addition he channel allows to set following values for triggering special actions <ul><li>EOD : triggers the "end of day" mode. The mower will switch in to the HOME mode and stay int this mode for the rest of the day</li><li ...
```
done

reyem commented:
@kaikreuzer ok, I think now the binding is in a quite good shape. Thank you very much for the extremely valuable feedback. That made the binding a bit slimmer and better 👍

reyem commented:
🎉 no 🦄 took over! 😆 

Many thanks to my first time user @DanielMalmgren for the reliable reports and eager testing.
Also thank you @kaikreuzer, @martinvw, @sjka and @Confectrician for your great reviews. They helped a lot to get the binding to its first merge state and also gave me the opportunity to learn useful things for further OH contributions or even other projects!

DanielMalmgren commented:
Thanks for good work @reyem ! I'm noticing that the review only took a little more than a year 🤣 

Anyway, what happens now? Does this mean the binding will appear as downloadable in paperui before 2.3 release?

reyem commented:
> Does this mean the binding will appear as downloadable in paperui before 2.3 release?

Yes, it is available in the 2.3.0-SNAPSHOT like all other bindings and will most likely make it in the final 2.3.0 Release.

> Anyway, what happens now?

Being part of the official bistro does not mean the effort stops. Next is extending the binding with features of the new and existing APIs of Robonect. On the TODO List is 

- Timer Support (at least enable & disable them)
- Find concepts to support push mode or mqtt integration to get rid of the polling or at least make it optional
- Channel for resetting errors
- Job Actions, once the OH/ESH concepts for actions in the rule engine is ready
- .. you name it

The good thing now is,

Anybody might create a PR and the PRs are likely much smaller with higher likelihood to get in in shorter time.

But please from now on let's discuss enhancements, future improvement whishes in the community form. The currently open thread for this is: https://community.openhab.org/t/robonect-binding-automower-wlan-module/27609/51

Thx for all your help
